pkg/deploy/controller/configchange/controller.go

```diff
@@ -42,20 +42,81 @@ func (c *DeploymentConfigChangeController) Handle(config *deployapi.DeploymentCo
 	}
 
 	if config.Status.LatestVersion == 0 {
-		_, _, abort, err := c.generateDeployment(config)
-		if err != nil {
-			if kerrors.IsConflict(err) {
-				return fatalError(fmt.Sprintf("deployment config %q updated since retrieval; aborting trigger: %v", deployutil.LabelForDeploymentConfig(config), err))
-			}
-			glog.V(4).Infof("Couldn't create initial deployment for deployment config %q: %v", deployutil.LabelForDeploymentConfig(config), err)
+		canTrigger, causes := c.canTrigger(config)
+		if !canTrigger {
+			// If we cannot trigger then we need to wait for the image change controller.
+			glog.V(5).Infof("Ignoring deployment config %q; template image needs to be resolved by the image change controller", deployutil.LabelForDeploymentConfig(config))
 			return nil
 		}
-		if !abort {
-			glog.V(4).Infof("Created initial deployment for deployment config %q", deployutil.LabelForDeploymentConfig(config))
-		}
+		return c.updateStatus(config, causes)
+	}
+
+	hasDiff, err := c.hasTemplateDiff(config)
+	if err != nil {
+		return err
+	}
+	if !hasDiff {
 		return nil
 	}
 
+	_, causes := c.canTrigger(config)
+
+	return c.updateStatus(config, causes)
+}
+
+// canTrigger is used by the config change controller to determine if the provided config can
+// trigger its initial deployment. The only requirement is set for image change trigger (ICT)
+// deployments - all of the ICTs need to have LastTriggedImage set which means that the image
+// change controller did its job. The second return argument helps in separating between config
+// change and image change causes.
+func (c *DeploymentConfigChangeController) canTrigger(config *deployapi.DeploymentConfig) (bool, []deployapi.DeploymentCause) {
+	ictCount, resolved := 0, 0
+	causes := []deployapi.DeploymentCause{}
```

Comment:
Use var causes [].... in situations like this - avoids an allocation if triggers is empty